src/System.Private.CoreLib/shared/System/Memory.cs

```diff
@@ -247,50 +239,91 @@ public Memory<T> Slice(int start)
-                }
-                else if (typeof(T) == typeof(char) && _object is string s)
-                {
-                    Debug.Assert(_length >= 0);
-                    // This is dangerous, returning a writable span for a string that should be immutable.
-                    // However, we need to handle the case where a ReadOnlyMemory<char> was created from a string
-                    // and then cast to a Memory<T>. Such a cast can only be done with unsafe or marshaling code,
-                    // in which case that's the dangerous operation performed by the dev, and we're just following
-                    // suit here to make it work as best as possible.
-                    return new Span<T>(ref Unsafe.As<char, T>(ref s.GetRawStringData()), s.Length).Slice(_index, _length);
-                }
-                else if (_object != null)
-                {
-                    return new Span<T>((T[])_object, _index, _length & RemoveFlagsBitMask);
-                }
-                else
+                // This property getter has special support for returning a mutable Span<char> that wraps
+                // an immutable String instance. This is obviously a dangerous feature and breaks type safety.
+                // However, we need to handle the case where a ReadOnlyMemory<char> was created from a string
+                // and then cast to a Memory<T>. Such a cast can only be done with unsafe or marshaling code,
+                // in which case that's the dangerous operation performed by the dev, and we're just following
+                // suit here to make it work as best as possible.
+
+                ref T refToReturn = ref Unsafe.AsRef<T>(null);
+                int lengthOfUnderlyingSpan = 0;
+
+                // Copy this field into a local so that it can't change out from under us mid-operation.
+
+                object tmpObject = _object;
+                if (tmpObject != null)
                 {
-                    return default;
+                    if (typeof(T) == typeof(char) && tmpObject.GetType() == typeof(string))
+                    {
+                        // Special-case string since it's the most common for ROM<char>.
+
+                        refToReturn = ref Unsafe.As<char, T>(ref Unsafe.As<string>(tmpObject).GetRawStringData());
+                        lengthOfUnderlyingSpan = Unsafe.As<string>(tmpObject).Length;
+                    }
+                    else if (RuntimeHelpers.ObjectHasComponentSize(tmpObject))
+                    {
+                        // We know the object is not null, it's not a string, and it is variable-length. The only
+                        // remaining option is for it to be a T[] (or a U[] which is blittable to T[], like int[]
+                        // and uint[]). Otherwise somebody used private reflection to set this field, and we're not
```

Comment:
> Otherwise somebody used private reflection

Or something like a union struct with explicit layout.

Reply:
Sure, but explicit layout is unsafe by construction.

```cs
  [StructLayout(LayoutKind.Explicit)]
  private ref struct MyStruct {
    [FieldOffset(0)]
    public ReadOnlySpan<char> TheString;

    [FieldOffset(8)]
    public int Length;
  }

  public static void Main()
  {
    MyStruct mys = default;
    mys.TheString = "Hello!";
    
    foreach (char i in mys.TheString) {
      Console.WriteLine(i);
    }

    mys.Length = 500;

    // below code will read well past the end of the original string
    foreach (char i in mys.TheString) {
      Console.WriteLine(i);
    }
  }
```

Reply:
Yup.  I was just pointing out it's not limited specifically to private reflection.